### PR TITLE
remove dev-dependencies resolution from the build

### PR DESCRIPTION
**Pull request by roxchkplusony**
Relates-to: python-poetry/poetry#2174

- [x] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

### CODE REVIEW

File: tests/fixtures/project_with_invalid_dev_deps/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+[tool.poetry]
```
Comment (abn) [inline tests/fixtures/project_with_invalid_dev_deps/pyproject.toml:1]: I would move this to the builders fixtures as this needs to be tested with api not with factory. Maybe name it `with_invalid_path_dep_dev` and add package to the source too so we can evaluate the built wheel.

Additionally, I would also recommend that we add a fixture for `with_invalid_path_dep_main` that can be used to ensure an error is raised if encountered.

File: poetry/core/masonry/api.py [hunk 2]
```diff
@@ -31,7 +31,7 @@ def get_requires_for_build_wheel(config_settings=None):
 
 
 def prepare_metadata_for_build_wheel(metadata_directory, config_settings=None):
-    poetry = Factory().create_poetry(Path(".").resolve())
+    poetry = Factory().create_poetry(Path(".").resolve(), with_dev=False)
```
Comment (abn) [inline poetry/core/masonry/api.py:34]: I do not think these tests need to change.

File: poetry/core/factory.py [hunk 3]
```diff
@@ -27,7 +27,7 @@ class Factory(object):
     Factory class to create various elements needed by Poetry.
     """
 
-    def create_poetry(self, cwd=None):  # type: (Optional[Path]) -> Poetry
+    def create_poetry(self, cwd=None, with_dev=True):  # type: (Optional[Path]) -> Poetry
```
Comment (abn) [inline poetry/core/factory.py:30]:
```suggestion
    def create_poetry(self, cwd=None, with_dev=True):  # type: (Optional[Path]. bool) -> Poetry
```

File: tests/test_factory.py [hunk 4]
```diff
@@ -199,3 +199,16 @@ def test_create_poetry_fails_on_invalid_configuration():
   - 'description' is a required property
 """
     assert expected == str(e.value)
+
+
+def test_create_poetry_with_invalid_dev_dependencies():
```
Comment (abn) [inline tests/test_factory.py:204]: The test cases here should be focussed on the flag `with_dev` directly, ie. re-use exostomg fixtured to check if development dependencies were loaded and not loaded if `True` and `False` respecively.

File: poetry/core/masonry/api.py [hunk 5]
```diff
@@ -52,14 +52,14 @@ def prepare_metadata_for_build_wheel(metadata_directory, config_settings=None):
 
 def build_wheel(wheel_directory, config_settings=None, metadata_directory=None):
     """Builds a wheel, places it in wheel_directory"""
-    poetry = Factory().create_poetry(Path(".").resolve())
+    poetry = Factory().create_poetry(Path(".").resolve(), with_dev=False)
```
Comment (abn) [inline poetry/core/masonry/api.py:55]: The verification of this shouldbe done by duplcating the existing test with the new fixtures added here. File content verification is not required, but just need to verify that a) wheel buidls when invalid dev dependency encountered and b) wheel builds will fail if these are encountered in the main category.

https://github.com/python-poetry/poetry-core/blob/0db07bc81b989c5c4933cf019fec7a4f93fbd744/tests/masonry/test_api.py#L46-L54

It might also be great to add an integration test case if easy enough. See the integration test suite at https://github.com/python-poetry/poetry-core/blob/master/tests/integration/test_pep517.py.

**Review by abn — CHANGES_REQUESTED** [review]

File: tests/fixtures/project_with_invalid_dev_deps/pyproject.toml (hunk 1 above)
Comment (roxchkplusony) [reply tests/fixtures/project_with_invalid_dev_deps/pyproject.toml:1]: Do you believe strongly that I should ONLY test in test_api and not in test_factory at all? I'll remove those tests if you don't want them, but I think they're doing something slightly different, something a little less e2e and applicative but more focused on the module itself.

File: tests/test_factory.py (hunk 4 above)
Comment (roxchkplusony) [reply tests/test_factory.py]: Tests added for that.

File: poetry/core/masonry/api.py (hunk 5 above)
Comment (roxchkplusony) [reply poetry/core/masonry/api.py:55]: I think for edge cases not meant to affect the build otherwise, it should stay in the lower level and not in the integration test. Let me know if you feel strongly about that though.

File: poetry/core/masonry/api.py (hunk 2 above)
Comment (roxchkplusony) [reply poetry/core/masonry/api.py:34]: Because this isn't test code I'm a little confused what you mean. Could you elaborate?

Comment (abn) [reply poetry/core/masonry/api.py:34]: Ignore me.

File: tests/fixtures/project_with_invalid_dev_deps/pyproject.toml (hunk 1 above)
Comment (abn) [reply tests/fixtures/project_with_invalid_dev_deps/pyproject.toml:1]: I believe you are testing the wrong things at the wrong level. The change is to ensure that when called via the PEP 517 api we ignore the dev dependnecies. The change in the factory code is to enable/disable inclusion of development dependencies and hence that is what should be tested, not if the "invalid" dependencies are not raising exceptions. This is what I have indicated in my other comment.

The suggestion to move the fixture is more for consistency as that is where the builder related fixtures are at. For verifying the flag, you can simply use an existing fixture here with both main and dev dependencies and parameterised test validating both flag states.

File: poetry/core/masonry/api.py (hunk 5 above)
Comment (abn) [reply poetry/core/masonry/api.py:55]: This is still a case, not an edge case per say. We are verifying that packages with invalid or non-existent dev dependnecies can still be built. This is not an edge case. 

Integration testing (`pip` or `pep517` call) is option for this, but api (masonry) level testing is mandatory.

File: tests/fixtures/project_with_invalid_dev_deps/pyproject.toml (hunk 1 above)
Comment (roxchkplusony) [reply tests/fixtures/project_with_invalid_dev_deps/pyproject.toml:1]: @abn I wrote new tests in test_api that I believe does what you are asking for. I also updated the test_factory tests which I believe also does what you are asking for. Did you get a chance to look at the new tests? Is your opinion still unchanged?

Comment (abn) [reply tests/fixtures/project_with_invalid_dev_deps/pyproject.toml:1]: Looks about right at first glance will review tomorrow. Thanks for the changes.

**Review by abn — CHANGES_REQUESTED** [review]
You seem to have inadvertantly moved `tests/json/__init__.py`. Additionally, it could be great if we could parametrise the tests reduce some duplication (optional).

Otherwise looks good.

**Comment by roxchkplusony** [comment]
Added back the __init__.py file I had inadvertently deleted. I don't see a great deduplication opportunity with the two test files, but let me know if your optional becomes mandatory. Otherwise, I hope I can get an approve.

**Review by abn — APPROVED** [review]

**Comment by orlevii** [comment]
Any plans to release this fix soon? 🙏
